// @tds/core-input/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));
var styled = _interopDefault(require('styled-components'));
var coreFeedbackIcon = require('@tds/core-feedback-icon');
var sharedAnimation = require('@tds/shared-animation');
var utilPropTypes = require('@tds/util-prop-types');
var InputFeedback = _interopDefault(require('@tds/core-input-feedback'));
var Box = _interopDefault(require('@tds/core-box'));
var Text = _interopDefault(require('@tds/core-text'));
var Paragraph = _interopDefault(require('@tds/core-paragraph'));
var sharedStyles = require('@tds/shared-styles');
var coreColours = require('@tds/core-colours');
var sharedTypography = require('@tds/shared-typography');
var utilHelpers = require('@tds/util-helpers');

var StyledIcon = styled.div.withConfig({
  displayName: "FeedbackIcon__StyledIcon",
  componentId: "yxz8rh-0"
})({
  lineHeight: 1
});

var renderIcon = function renderIcon(feedback) {
  if (feedback === 'success') {
    return React__default.createElement(coreFeedbackIcon.NotificationSuccess, {
      copy: {
        a11yText: 'The value of this input field is valid.'
      }
    });
  }

  if (feedback === 'error') {
    return React__default.createElement(coreFeedbackIcon.NotificationError, {
      copy: {
        a11yText: 'The value of this input field is invalid.'
      }
    });
  }

  return null;
};

var FeedbackIcon = function FeedbackIcon(_ref) {
  var showIcon = _ref.showIcon,
      feedback = _ref.feedback;
  return React__default.createElement(sharedAnimation.Fade, {
    timeout: 100,
    in: showIcon,
    mountOnEnter: true,
    unmountOnExit: true
  }, function () {
    return React__default.createElement(StyledIcon, null, renderIcon(feedback));
  });
};

FeedbackIcon.propTypes = {
  showIcon: PropTypes.bool.isRequired,
  feedback: PropTypes.oneOf(['success', 'error'])
};
FeedbackIcon.defaultProps = {
  feedback: undefined
};

function _taggedTemplateLiteral(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
}

var taggedTemplateLiteral = _taggedTemplateLiteral;

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var _extends_1 = createCommonjsModule(function (module) {
function _extends() {
  module.exports = _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

module.exports = _extends;
});

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

var arrayWithHoles = _arrayWithHoles;

function _iterableToArrayLimit(arr, i) {
  if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) {
    return;
  }

  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

var iterableToArrayLimit = _iterableToArrayLimit;

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance");
}

var nonIterableRest = _nonIterableRest;

function _slicedToArray(arr, i) {
  return arrayWithHoles(arr) || iterableToArrayLimit(arr, i) || nonIterableRest();
}

var slicedToArray = _slicedToArray;

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

var objectWithoutPropertiesLoose = _objectWithoutPropertiesLoose;

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};
  var target = objectWithoutPropertiesLoose(source, excluded);
  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

var objectWithoutProperties = _objectWithoutProperties;

function find(t,n,r){if("function"==typeof Array.prototype.find)return t.find(n,r);r=r||this;var f=t.length;if("function"!=typeof n)throw new TypeError(n+" is not a function");for(var o=0;o<f;o++)if(n.call(r,t[o],o,t))return t[o]}

var sanitize = function sanitize(text) {
  return text.toString().toLowerCase().replace(/ /g, '-').replace(/[^a-zA-Z0-9-]/g, '');
};

var generateId = function generateId() {
  for (var _len = arguments.length, choices = new Array(_len), _key = 0; _key < _len; _key++) {
    choices[_key] = arguments[_key];
  }

  var id = sanitize(find(choices, function (choice) {
    return choice;
  }));
  return {
    identity: function identity() {
      return id;
    },
    postfix: function postfix(value) {
      return "".concat(id, "_").concat(sanitize(value));
    }
  };
};

function _templateObject() {
  var data = taggedTemplateLiteral(["", ""]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var StyledLabelContainer = styled(Box)({
  alignItems: 'center'
});
var StyledInput = styled.input.withConfig({
  displayName: "Input__StyledInput",
  componentId: "sc-8rmijy-0"
})({
  width: '100%',
  margin: 0,
  outline: 0,
  textOverflow: 'ellipsis',
  borderColor: coreColours.colorGreyShuttle,
  '&::placeholder': {
    font: 'inherit',
    letterSpacing: 'inherit',
    lineHeight: 'inherit',
    color: coreColours.colorGreyShuttle
  }
}, sharedStyles.borders.thin, sharedStyles.borders.rounded, sharedStyles.forms.font, sharedTypography.medium, sharedTypography.mediumFont, sharedTypography.color, function (_ref) {
  var withFeedbackIcon = _ref.withFeedbackIcon;
  return {
    '&::-webkit-inner-spin-button, &::-webkit-outer-spin-button': {
      appearance: 'none',
      margin: 0
    },
    '-moz-appearance': 'textfield',
    minHeight: sharedStyles.forms.inputHeight.height,
    maxHeight: sharedStyles.forms.inputHeight.height,
    padding: withFeedbackIcon ? '0.5rem 3rem 0.5rem 1rem' : '0.5rem 1rem'
  };
}, {
  '&:focus': {
    borderColor: 'transparent',
    boxShadow: "0 0 4px 1px ".concat(coreColours.colorGreyShuttle),
    backgroundColor: coreColours.colorWhite
  }
}, function (_ref2) {
  var feedback = _ref2.feedback;
  var borderColor;

  if (feedback === 'success') {
    borderColor = coreColours.colorPrimary;
  } else if (feedback === 'error') {
    borderColor = coreColours.colorCardinal;
  }

  return {
    '&:not(:focus)': {
      borderColor: borderColor
    }
  };
}, function (_ref3) {
  var disabled = _ref3.disabled;

  if (disabled) {
    return {
      backgroundColor: coreColours.colorGreyAthens,
      borderColor: 'transparent',
      '&:not(:focus)': {
        borderColor: 'transparent'
      }
    };
  }

  return {};
});
var StyledFeedbackIcon = styled.div.withConfig({
  displayName: "Input__StyledFeedbackIcon",
  componentId: "sc-8rmijy-1"
})({
  right: '1rem'
}, sharedStyles.position.absolute, sharedStyles.position.centerVertically);

var showFeedbackIcon = function showFeedbackIcon(feedback) {
  return feedback === 'error' || feedback === 'success';
};

var renderHint = function renderHint(hint, Component, id) {
  return React__default.createElement(Component, {
    id: id,
    size: "small"
  }, hint);
};

var renderError = function renderError(error, errorId) {
  return React__default.createElement(InputFeedback, {
    id: errorId,
    feedback: "error"
  }, React__default.createElement(Paragraph, {
    size: "small"
  }, error));
};

var renderLabel = function renderLabel(id, label, hint, hintPosition, hintId, tooltip) {
  return React__default.createElement("div", null, React__default.createElement(Box, {
    inline: true,
    between: 2
  }, React__default.createElement("label", {
    htmlFor: id || generateId(label).identity()
  }, React__default.createElement(StyledLabelContainer, {
    inline: true,
    tag: "span",
    between: 2
  }, React__default.createElement(Text, {
    size: "medium",
    bold: true
  }, label), hint && hintPosition === 'inline' && renderHint(hint, Text, hintId))), tooltip && React__default.cloneElement(tooltip, {
    connectedFieldLabel: label
  })), hint && hintPosition === 'below' && renderHint(hint, Paragraph, hintId));
};

var renderHelper = function renderHelper(helper, helperId, feedback, value) {
  if (typeof helper === 'function') {
    return React__default.createElement("div", {
      id: helperId
    }, React__default.createElement(Text, {
      size: "small"
    }, helper(feedback, value)));
  }

  return React__default.createElement(InputFeedback, {
    id: helperId,
    feedback: feedback
  }, React__default.createElement(Text, {
    size: "small"
  }, helper));
};
/**
 * @version ./package.json
 */


var Input = React__default.forwardRef(function (_ref4, ref) {
  var id = _ref4.id,
      value = _ref4.value,
      type = _ref4.type,
      label = _ref4.label,
      hint = _ref4.hint,
      hintPosition = _ref4.hintPosition,
      feedback = _ref4.feedback,
      error = _ref4.error,
      helper = _ref4.helper,
      tooltip = _ref4.tooltip,
      rest = objectWithoutProperties(_ref4, ["id", "value", "type", "label", "hint", "hintPosition", "feedback", "error", "helper", "tooltip"]);

  var _useState = React.useState(false),
      _useState2 = slicedToArray(_useState, 2),
      isFocused = _useState2[0],
      setIsFocused = _useState2[1];

  var fieldId = generateId(id, rest.name, label);
  var errorId = error && fieldId.postfix('error-message');
  var helperId = helper && fieldId.postfix('helper');
  var hintId = hint && hintPosition === 'below' && fieldId.postfix('hint') || undefined;

  var handleFocus = function handleFocus(e) {
    setIsFocused(true);

    if (rest.onFocus) {
      rest.onFocus(e);
    }
  };

  var handleBlur = function handleBlur(e) {
    setIsFocused(false);

    if (rest.onBlur) {
      rest.onBlur(e);
    }
  };

  var handleKeyDown = function handleKeyDown(e) {
    /**
     * this is a workaround for a bug in chrome that moves
     * the cursor into a wrong position if prepended with a space
     */
    if (type === 'email' && e.key === ' ') {
      e.preventDefault();
    }

    if (rest.onKeyDown) {
      rest.onKeyDown(e);
    }
  };

  return React__default.createElement(Box, {
    between: 2
  }, renderLabel(fieldId.identity(), label, hint, hintPosition, hintId, tooltip), helper && renderHelper(helper, helperId, feedback, value), error && renderError(error, errorId), React__default.createElement(_StyledDiv, {
    _css: {
      position: 'relative'
    }
  }, React__default.createElement(StyledInput, _extends_1({}, utilHelpers.safeRest(rest), {
    type: type,
    ref: ref,
    id: fieldId.identity(),
    value: value,
    feedback: feedback,
    "aria-invalid": feedback === 'error',
    "aria-describedby": errorId || hintId || helperId || undefined,
    onFocus: handleFocus,
    onBlur: handleBlur,
    onKeyDown: handleKeyDown
  })), !rest.disabled && React__default.createElement(StyledFeedbackIcon, null, React__default.createElement(FeedbackIcon, {
    showIcon: showFeedbackIcon(feedback) && !isFocused,
    feedback: feedback
  }))));
});
Input.displayName = 'Input';
Input.propTypes = {
  /**
   * The id.
   */
  id: PropTypes.string,

  /**
   * The HTML5 type of the input field.
   */
  type: PropTypes.oneOf(['text', 'number', 'password', 'email', 'search', 'tel', 'url']),

  /**
   * The label.
   */
  label: PropTypes.string.isRequired,

  /**
   * Clarify the expected input.
   */
  hint: PropTypes.string,

  /**
   * Position of the `hint` relative to `label`.
   */
  hintPosition: PropTypes.oneOf(['inline', 'below']),

  /**
   * Use `value` for controlled Inputs. For uncontrolled Inputs, use React's built-in `defaultValue` prop.
   * See examples below for more details.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(['success', 'error']),

  /**
   * An error message. Should be limited to text and links. See usage criteria for more details.
   */
  error: PropTypes.node,

  /**
   * A detailed explanation of the input expected by a form field. Can be text,
   * other components, or HTML elements.
   *
   * If a function is provided, it must return an `InputFeedback`. The function will be
   * invoked with arguments below.
   *
   * **Deprecation:** This is not a recommended pattern and will be removed in a future release.
   *                  Use `hint` with `hintPosition` value of `below`.
   *
   * @param {String} feedback The Input's current feedback state.
   * @param {String} value The Input's current value.
   */
  helper: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),

  /**
   * A `Tooltip`
   */
  tooltip: utilPropTypes.componentWithName('Tooltip', true)
};
Input.defaultProps = {
  id: undefined,
  type: 'text',
  hint: undefined,
  hintPosition: 'inline',
  value: undefined,
  feedback: undefined,
  error: undefined,
  tooltip: undefined,
  helper: undefined
};

var _StyledDiv = styled("div")(_templateObject(), function (p) {
  return p._css;
});

exports.FeedbackIcon = FeedbackIcon;
exports.default = Input;
