// @tds/core-text-button/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var PropTypes = _interopDefault(require('prop-types'));
var styled = _interopDefault(require('styled-components'));
var utilPropTypes = require('@tds/util-prop-types');
var coreColours = require('@tds/core-colours');
var sharedTypography = require('@tds/shared-typography');
var utilHelpers = require('@tds/util-helpers');

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

var objectWithoutPropertiesLoose = _objectWithoutPropertiesLoose;

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};
  var target = objectWithoutPropertiesLoose(source, excluded);
  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

var objectWithoutProperties = _objectWithoutProperties;

var baseButton = {
  boxSizing: 'border-box',
  padding: '0.25rem 0rem',
  cursor: 'pointer',
  background: coreColours.colorWhite,
  transition: 'all 0.2s ease-in-out',
  position: 'relative',
  borderRadius: '0.1875rem',
  color: coreColours.colorGreyShark,
  textDecoration: 'underline',
  borderStyle: 'none',
  '&:hover': {
    textDecoration: 'none'
  },
  '&:active': {
    background: coreColours.colorGreyGainsboro,
    boxShadow: "0 0 0 0.125rem ".concat(coreColours.colorGreyGainsboro),
    textDecoration: 'underline'
  },
  '&:focus': {
    outline: 'none !important',
    boxShadow: "0 0 0 0.125rem ".concat(coreColours.colorGreyRaven)
  },
  '@media (prefers-reduced-motion: reduce)': {
    transition: 'none !important'
  }
};
var StyledTextButton = styled.button.withConfig({
  displayName: "TextButton__StyledTextButton",
  componentId: "i04c5a-0"
})(baseButton, sharedTypography.medium);
/**
 * @version ./package.json
 */

var TextButton = function TextButton(_ref) {
  var children = _ref.children,
      rest = objectWithoutProperties(_ref, ["children"]);

  return React.createElement(StyledTextButton, utilHelpers.safeRest(rest), children);
};

TextButton.propTypes = {
  /**
   * The label. It can include the `A11yContent` component or strings.
   */
  children: utilPropTypes.or([PropTypes.string, utilPropTypes.componentWithName('A11yContent')]).isRequired
};

exports.StyledTextButton = StyledTextButton;
exports.default = TextButton;
