// @tds/core-select/dist/index.cjs.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var styled = _interopDefault(require('styled-components'));
var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));
var Box = _interopDefault(require('@tds/core-box'));
var coreInteractiveIcon = require('@tds/core-interactive-icon');
var coreInput = require('@tds/core-input');
var Text = _interopDefault(require('@tds/core-text'));
var Paragraph = _interopDefault(require('@tds/core-paragraph'));
var utilPropTypes = require('@tds/util-prop-types');
var InputFeedback = _interopDefault(require('@tds/core-input-feedback'));
var sharedStyles = require('@tds/shared-styles');
var coreColours = require('@tds/core-colours');
var sharedTypography = require('@tds/shared-typography');
var utilHelpers = require('@tds/util-helpers');

function _taggedTemplateLiteral(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
}

var taggedTemplateLiteral = _taggedTemplateLiteral;

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

var arrayWithHoles = _arrayWithHoles;

function _iterableToArrayLimit(arr, i) {
  if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) {
    return;
  }

  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

var iterableToArrayLimit = _iterableToArrayLimit;

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance");
}

var nonIterableRest = _nonIterableRest;

function _slicedToArray(arr, i) {
  return arrayWithHoles(arr) || iterableToArrayLimit(arr, i) || nonIterableRest();
}

var slicedToArray = _slicedToArray;

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;

  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }

  return target;
}

var objectWithoutPropertiesLoose = _objectWithoutPropertiesLoose;

function _objectWithoutProperties(source, excluded) {
  if (source == null) return {};
  var target = objectWithoutPropertiesLoose(source, excluded);
  var key, i;

  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

    for (i = 0; i < sourceSymbolKeys.length; i++) {
      key = sourceSymbolKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
      target[key] = source[key];
    }
  }

  return target;
}

var objectWithoutProperties = _objectWithoutProperties;

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var _extends_1 = createCommonjsModule(function (module) {
function _extends() {
  module.exports = _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

module.exports = _extends;
});

function find(t,n,r){if("function"==typeof Array.prototype.find)return t.find(n,r);r=r||this;var f=t.length;if("function"!=typeof n)throw new TypeError(n+" is not a function");for(var o=0;o<f;o++)if(n.call(r,t[o],o,t))return t[o]}

var sanitize = function sanitize(text) {
  return text.toString().toLowerCase().replace(/ /g, '-').replace(/[^a-zA-Z0-9-]/g, '');
};

var generateId = function generateId() {
  for (var _len = arguments.length, choices = new Array(_len), _key = 0; _key < _len; _key++) {
    choices[_key] = arguments[_key];
  }

  var id = sanitize(find(choices, function (choice) {
    return choice;
  }));
  return {
    identity: function identity() {
      return id;
    },
    postfix: function postfix(value) {
      return "".concat(id, "_").concat(sanitize(value));
    }
  };
};

function _templateObject() {
  var data = taggedTemplateLiteral(["", ""]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var SelectWrapper = styled.div.withConfig({
  displayName: "Select__SelectWrapper",
  componentId: "sc-1kajzpi-0"
})(_extends_1({
  backgroundColor: coreColours.colorWhite
}, sharedStyles.position.relative));
var StyledSelect = styled.select.withConfig({
  displayName: "Select__StyledSelect",
  componentId: "sc-1kajzpi-1"
})(_extends_1({}, sharedStyles.forms.inputHeight, {
  appearance: 'none',
  '&::-ms-expand': {
    display: 'none'
  },
  width: '100%',
  margin: 0,
  outline: 0,
  textOverflow: 'ellipsis',
  borderColor: coreColours.colorGreyShuttle,
  backgroundColor: coreColours.colorWhite,
  '&::placeholder': {
    font: 'inherit',
    letterSpacing: 'inherit',
    lineHeight: 'inherit',
    color: coreColours.colorGreyShuttle
  }
}), function (_ref) {
  var showFeedbackIcon = _ref.showFeedbackIcon;
  return {
    padding: "0.5rem ".concat(showFeedbackIcon ? '4' : '3', "rem 0.5rem 1rem")
  };
}, sharedStyles.borders.thin, sharedStyles.borders.rounded, sharedStyles.forms.font, sharedTypography.medium, sharedTypography.mediumFont, sharedTypography.color, function (_ref2) {
  var withFeedbackIcon = _ref2.withFeedbackIcon;
  return {
    '&::-webkit-inner-spin-button, &::-webkit-outer-spin-button': {
      appearance: 'none',
      margin: 0
    },
    minHeight: sharedStyles.forms.inputHeight.height,
    maxHeight: sharedStyles.forms.inputHeight.height,
    padding: withFeedbackIcon ? '0.5rem 4rem 0.5rem 1rem' : '0.5rem 3rem 0.5rem 1rem'
  };
}, {
  '&:focus': {
    borderColor: 'transparent',
    boxShadow: "0 0 4px 1px ".concat(coreColours.colorGreyShuttle),
    backgroundColor: coreColours.colorWhite
  }
}, function (_ref3) {
  var feedback = _ref3.feedback;
  var borderColor;

  if (feedback === 'success') {
    borderColor = coreColours.colorPrimary;
  } else if (feedback === 'error') {
    borderColor = coreColours.colorCardinal;
  }

  return {
    '&:not(:focus)': {
      borderColor: borderColor
    }
  };
}, function (_ref4) {
  var disabled = _ref4.disabled;

  if (disabled) {
    return {
      backgroundColor: coreColours.colorGreyAthens,
      borderColor: 'transparent',
      '&:not(:focus)': {
        borderColor: 'transparent'
      }
    };
  }

  return {};
});
var IconWrapper = styled(Box)(_extends_1({}, sharedStyles.position.absolute, {
  top: '50%',
  transform: 'translateY(-50%)',
  right: '1rem',
  pointerEvents: 'none',
  display: 'flex',
  alignItems: 'center',
  borderColor: 'red'
}));
var IconLineFix = styled.div.withConfig({
  displayName: "Select__IconLineFix",
  componentId: "sc-1kajzpi-2"
})({
  lineHeight: 1
});
var StyledLabelContainer = styled(Box)({
  alignItems: 'center'
});

var showFeedbackIcon = function showFeedbackIcon(feedback) {
  return feedback === 'error' || feedback === 'success';
};

var renderHint = function renderHint(hint, Component, id) {
  return React__default.createElement(Component, {
    id: id,
    size: "small"
  }, hint);
};

var renderError = function renderError(error, errorId) {
  return React__default.createElement(InputFeedback, {
    id: errorId,
    feedback: "error"
  }, React__default.createElement(Paragraph, {
    size: "small"
  }, error));
};

var renderLabel = function renderLabel(id, label, hint, hintPosition, hintId, tooltip) {
  return React__default.createElement("div", null, React__default.createElement(Box, {
    inline: true,
    between: 2
  }, React__default.createElement("label", {
    htmlFor: id || generateId(label).identity()
  }, React__default.createElement(StyledLabelContainer, {
    inline: true,
    tag: "span",
    between: 2
  }, React__default.createElement(Text, {
    size: "medium",
    bold: true,
    "data-testid": "selectLabel"
  }, label), hint && hintPosition === 'inline' && renderHint(hint, Text, hintId))), tooltip && React__default.cloneElement(tooltip, {
    connectedFieldLabel: label
  })), hint && hintPosition === 'below' && renderHint(hint, Paragraph, hintId));
};

var renderHelper = function renderHelper(helper, helperId, feedback, value) {
  if (typeof helper === 'function') {
    return React__default.createElement("div", {
      id: helperId
    }, React__default.createElement(Text, {
      size: "small"
    }, helper(feedback, value)));
  }

  return React__default.createElement(InputFeedback, {
    id: helperId,
    feedback: feedback
  }, React__default.createElement(Text, {
    size: "small"
  }, helper));
};

var renderOptions = function renderOptions(opts) {
  return opts.map(function (_ref5) {
    var text = _ref5.text,
        optValue = _ref5.value,
        options = _ref5.options;

    if (options) {
      return React__default.createElement("optgroup", {
        label: text,
        key: text
      }, renderOptions(options));
    }

    return React__default.createElement("option", {
      key: optValue,
      value: optValue
    }, text);
  });
};
/**
 * @version ./package.json
 */


var Select = React__default.forwardRef(function (_ref6, ref) {
  var id = _ref6.id,
      options = _ref6.options,
      placeholder = _ref6.placeholder,
      label = _ref6.label,
      hint = _ref6.hint,
      hintPosition = _ref6.hintPosition,
      value = _ref6.value,
      defaultValue = _ref6.defaultValue,
      feedback = _ref6.feedback,
      error = _ref6.error,
      helper = _ref6.helper,
      tooltip = _ref6.tooltip,
      rest = objectWithoutProperties(_ref6, ["id", "options", "placeholder", "label", "hint", "hintPosition", "value", "defaultValue", "feedback", "error", "helper", "tooltip"]);

  var _useState = React.useState(false),
      _useState2 = slicedToArray(_useState, 2),
      isFocused = _useState2[0],
      setIsFocused = _useState2[1];

  var fieldId = generateId(id, rest.name, label);
  var errorId = error && feedback === 'error' && fieldId.postfix('error-message');
  var helperId = helper && fieldId.postfix('helper');
  var hintId = hint && fieldId.postfix('hint') || undefined;

  var handleFocus = function handleFocus(e) {
    setIsFocused(true);

    if (rest.onFocus) {
      rest.onFocus(e);
    }
  };

  var handleBlur = function handleBlur(e) {
    setIsFocused(false);

    if (rest.onBlur) {
      rest.onBlur(e);
    }
  };

  return React__default.createElement(Box, {
    between: 2
  }, renderLabel(fieldId.identity(), label, hint, hintPosition, hintId, tooltip), helper && renderHelper(helper, helperId, feedback, value), error && feedback === 'error' && renderError(error, errorId), React__default.createElement(_StyledDiv, {
    _css: {
      position: 'relative'
    }
  }, React__default.createElement(SelectWrapper, null, React__default.createElement(StyledSelect, _extends_1({}, utilHelpers.safeRest(rest), {
    ref: ref,
    id: fieldId.identity(),
    value: value,
    feedback: feedback,
    "aria-invalid": feedback === 'error',
    "aria-describedby": errorId || hintId || helperId || undefined,
    onFocus: handleFocus,
    onBlur: handleBlur,
    defaultValue: value !== undefined ? undefined : defaultValue
  }), placeholder && React__default.createElement("option", {
    value: "",
    disabled: true
  }, placeholder), renderOptions(options)), !rest.disabled && React__default.createElement(IconWrapper, {
    inline: true,
    between: 1
  }, React__default.createElement(coreInput.FeedbackIcon, {
    showIcon: showFeedbackIcon(feedback) && !isFocused,
    feedback: feedback
  }), React__default.createElement(IconLineFix, null, React__default.createElement(coreInteractiveIcon.CaretDown, {
    variant: feedback === 'error' ? 'error' : undefined,
    size: 16
  }))))));
});
Select.displayName = 'Select';
Select.propTypes = {
  /**
   * The id.
   */
  id: PropTypes.string,

  /**
   * The selectable options.
   *
   * Updated in v3.2.0 to support option groups (see docs).
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    text: PropTypes.string,
    value: PropTypes.string,
    options: PropTypes.array
  })).isRequired,

  /**
   * The label.
   */
  label: PropTypes.string.isRequired,

  /**
   * Clarify the expected input.
   */
  hint: PropTypes.string,

  /**
   * Position of the `hint` relative to `label`.
   *
   * @since 3.1.0
   */
  hintPosition: PropTypes.oneOf(['inline', 'below']),

  /**
   * Show a un-selectable initial option.
   */
  placeholder: PropTypes.string,

  /**
   * Use `value` for controlled Select. For uncontrolled Select, use React's built-in `defaultValue` prop.
   * See examples below for more details.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
   * The default value. Set this prop to set the inital selected option.
   *
   * Only one of defaultValue or value should be set / mutated. `value` will override `defaultValue`.
   */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(['success', 'error']),

  /**
   * An error message. Either an error or a helper should be used, not both.
   */
  error: PropTypes.string,

  /**
   * A detailed explanation of the input expected by a form field. Can be text,
   * other components, or HTML elements. Only appears when `feedback` is set to `error`.
   *
   * If a function is provided, it must return an `InputFeedback` component. The function will be
   * invoked with the following arguments.
   *
   * @param {String} feedback The input's current feedback state.
   * @param {String} value The input's current value.
   */
  helper: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),

  /**
   * A `Tooltip`
   */
  tooltip: utilPropTypes.componentWithName('Tooltip', true)
};
Select.defaultProps = {
  id: undefined,
  hint: undefined,
  hintPosition: 'inline',
  placeholder: undefined,
  value: undefined,
  defaultValue: '',
  feedback: undefined,
  error: undefined,
  helper: undefined,
  tooltip: undefined
};

var _StyledDiv = styled("div")(_templateObject(), function (p) {
  return p._css;
});

module.exports = Select;
