// @tds/core-colours/dist/index.es.js
// colours
var colorShark = '#2a2c2e';
var colorAthensGrey = '#f7f7f8';
var colorShuttleGrey = '#54595f';
var colorGainsboro = '#d8d8d8';
var colorDarkRaja = '#8C5415';
var colorRaja = '#FACA69';
var colorLightRaja = '#FFF9EE'; // brand

var colorTelusGreen = '#6c0';
var colorTelusPurple = '#4b286d';
var colorAccessibleGreen = '#2B8000';
var colorWhite = '#fff'; // grey

var colorGreyShark = '#2a2c2e';
var colorGreyShuttle = '#54595f';
var colorGreyRaven = '#71757b';
var colorGreyGainsboro = '#d8d8d8';
var colorGreyAthens = '#f7f7f8'; // notification

var colorLavenderBlush = '#fff6f8';
var colorPanache = '#f4f9f2';
var colorWhiteLilac = '#f2eff4';
var colorCardinal = '#c12335';
var colorRajahDark = '#8C5415';
var colorRajah = '#FACA69';
var colorRajahLight = '#FFF9EE'; // typography

var colorLink = colorGreyShuttle;
var colorText = colorGreyShark; // tokens

var colorPrimary = colorAccessibleGreen;
var colorSecondary = colorTelusPurple; // icons

var colorIconPrimary = colorPrimary;
var colorIconSecondary = colorSecondary;
var colorIconDisabled = colorGreyShuttle;
var colorIconError = colorCardinal;

export { colorAccessibleGreen, colorAthensGrey, colorCardinal, colorDarkRaja, colorGainsboro, colorGreyAthens, colorGreyGainsboro, colorGreyRaven, colorGreyShark, colorGreyShuttle, colorIconDisabled, colorIconError, colorIconPrimary, colorIconSecondary, colorLavenderBlush, colorLightRaja, colorLink, colorPanache, colorPrimary, colorRaja, colorRajah, colorRajahDark, colorRajahLight, colorSecondary, colorShark, colorShuttleGrey, colorTelusGreen, colorTelusPurple, colorText, colorWhite, colorWhiteLilac };
